[nnc] Fix dtype promotion involving scalars

bertmaher opened the pull request:
Fixes https://github.com/pytorch/vision/issues/4315

Type promotion involving tensors and scalars in PyTorch is super complicated, but for the fuser's purpose we can reduce it to a fairly simple procedure.  Basically, scalars will perform int->fp promotion, but not fp->higher precision fp (https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc).  Actually detecting a scalar in the tensor expression is a bit dicey, but I think the predicate I came up with is the right one.

As a bonus, remainder() is just kind of broken in some subtle way (which I think was swept under the covers by performing it in higher precision).  So I'm disabling it for the time being.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64002

Differential Revision: [D30566979](https://our.internmc.facebook.com/intern/diff/D30566979)

bertmaher committed:
[nnc] Fix dtype promotion involving scalars

[ghstack-poisoned]

navahgar reviewed (approved):
Mostly LGTM. 

I have one question below. But I don't want to delay landing this just for that.

bertmaher committed:
Update on "[nnc] Fix dtype promotion involving scalars"

Fixes https://github.com/pytorch/vision/issues/4315

Type promotion involving tensors and scalars in PyTorch is super complicated, but for the fuser's purpose we can reduce it to a fairly simple procedure.  Basically, scalars will perform int->fp promotion, but not fp->higher precision fp (https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc).  Actually detecting a scalar in the tensor expression is a bit dicey, but I think the predicate I came up with is the right one.

As a bonus, remainder() is just kind of broken in some subtle way (which I think was swept under the covers by performing it in higher precision).  So I'm disabling it for the time being.


Differential Revision: [D30566979](https://our.internmc.facebook.com/intern/diff/D30566979)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Fix dtype promotion involving scalars"

Fixes https://github.com/pytorch/vision/issues/4315

Type promotion involving tensors and scalars in PyTorch is super complicated, but for the fuser's purpose we can reduce it to a fairly simple procedure.  Basically, scalars will perform int->fp promotion, but not fp->higher precision fp (https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc).  Actually detecting a scalar in the tensor expression is a bit dicey, but I think the predicate I came up with is the right one.

As a bonus, remainder() is just kind of broken in some subtle way (which I think was swept under the covers by performing it in higher precision).  So I'm disabling it for the time being.


Differential Revision: [D30566979](https://our.internmc.facebook.com/intern/diff/D30566979)

[ghstack-poisoned]
